b_pull_Landsat_SRST_poi/py/gee_functions.py

```diff
@@ -0,0 +1,674 @@
+  and surface temperature products
+
+  Args:
+      image: one ee.Image of an ee.ImageCollection
+
+  Returns:
+      ee.Image with band values overwritten by scaling factors
+  """
+  opticalBands = image.select("SR_B.").multiply(0.0000275).add(-0.2)
+  thermalBands = image.select("ST_B.*").multiply(0.00341802).add(149.0)
+  return image.addBands(opticalBands, None, True).addBands(thermalBands, None,True)
+
+
+def dp_buff(image):
+  """ Buffer ee.FeatureCollection sites from csv_to_eeFeat by user-specified radius
+
+  Args:
+      image: ee.Image of an ee.ImageCollection
+
+  Returns:
+      ee.FeatureCollection of polygons resulting from buffered points
+  """
+  return image.buffer(ee.Number.parse(str(buffer)))
+
+
+def add_rad_mask(image):
+  """Mask out all pixels that are radiometrically saturated using the QA_RADSAT
+  QA band.
+
+  Args:
+      image: ee.Image of an ee.ImageCollection
+
+  Returns:
+      ee.Image with additional band called "radsat", where pixels with a value 
+      of 0 are saturated for at least one SR band and a value of 1 is not saturated
+  """
+  #grab the radsat band
+  satQA = image.select("radsat_qa")
+  # all must be non-saturated per pixel
+  satMask = satQA.eq(0).rename("radsat")
+  return image.addBands(satMask).updateMask(satMask)
+
+
+def cf_mask(image):
+  """Masks any pixels obstructed by clouds and snow/ice
+
+  Args:
+      image: ee.Image of an ee.ImageCollection
+
+  Returns:
+      ee.Image with additional band called "cfmask", where pixels are given values
+      based on the QA_PIXEL band informaiton. Generally speaking, 0 is clear, values 
+      greater than 0 are obstructed by clouds and/or snow/ice
+  """
+  #grab just the pixel_qa info
+  qa = image.select("pixel_qa")
+  cloudqa = (qa.bitwiseAnd(1 << 1).rename("cfmask") #dialated clouds value 1
+    .where(qa.bitwiseAnd(1 << 3), ee.Image(2)) # clouds value 2
+    .where(qa.bitwiseAnd(1 << 4), ee.Image(3)) # cloud shadows value 3
+    .where(qa.bitwiseAnd(1 << 5), ee.Image(4))) # snow value 4
+  return image.addBands(cloudqa)
+
+
+def sr_cloud_mask(image):
+  """Masks any pixles in Landsat 4-7 that are contaminated by the inputs of 
+  the atmospheric processing steps
+
+  Args:
+      image: ee.Image of an ee.ImageCollection
+
+  Returns:
+      ee.Image with additional band called "sr_cloud", where pixels are given values
+      based on the SR_CLOUD_QA band informaiton. Generally speaking, 0 is clear, values
+      greater than 0 are obstructed by clouds and/or snow/ice specifically from atmospheric
+      processing steps
+  """
+  srCloudQA = image.select("cloud_qa")
+  srMask = (srCloudQA.bitwiseAnd(1 << 1).rename("sr_cloud") # cloud
+    .where(srCloudQA.bitwiseAnd(1 << 2), ee.Image(2)) # cloud shadow
+    .where(srCloudQA.bitwiseAnd(1 << 3), ee.Image(3)) # adjacent to cloud
+    .where(srCloudQA.bitwiseAnd(1 << 4), ee.Image(4))) # snow/ice
+  return image.addBands(srMask)
+
```

Comment:
What's the difference in cloud classification between cfmask and cloud_qa? I have not been working on C2 data as much these days...

Reply:
the cloud qa is only for ls 4-7, and my understanding is that tracks clouds/potential issues in the products used to create the surface reflectance product. This is the mask that I noted in our optical theory meeting that is relatively aggressive at masking pixels in 4-7 and that we're going to do some internal comparisons to see if it is necessary to use as a mask, or if we should use it as a pCount summary statistic for QAQC filtering.

Reply:
Thank you for explaining it again. My memory is really bad nowadays. So the cfmask is identical to those in the C1 data? does the cloud_qa also indicate the problems associated with the erroneous aerosol correction that comes with C2 data?